### DISCUSSION TITLE
is it possible to show scripts based on focused window?

### DISCUSSION THREAD

**Discussion by yangkennyk** [johnlindquist/kit#1131] [Q&A] [answered]
I can't seem to find info on if this is possible.

Basically I'd like to create command bar per app/window.

So activating kit on vscode would show my vscode scripts, chrome would show my chrome related scripts, etc.

Thanks!

**Comment by johnlindquist** [comment]
@yangkennyk What's your level of experience with programming? Here's a great package which you can use and put at the front of your script and then display actions/scripts based on the app. Let me know if you'd  like a demo.

https://www.npmjs.com/package/frontmost-app

**Comment by yangkennyk** [comment]
@johnlindquist I'm a developer so hopefully I can figure it out 😅. 

although the first thing that comes to mind is how would I actually exclude a script from showing up when I bring up script kit? would I have a script kit file that simply acts as a conditional check to see if the current app = X and if so, then dynamically import the relevant script files?

**Comment by johnlindquist** [comment, marked as answer]
@yangkennyk something like this would do the trick. I'm basing this on custom `// App:` metadata.

> Not sure of your familiarity with TypeScript, but you can remove the `any` (or if using Typescript, extend/create types)

```js
// Name: Scripts Based on Focused App
// Shortcut: cmd 0

import "@johnlindquist/kit"

let frontmost = await npm("frontmost-app")

let app = await frontmost()

let scripts = await getScripts()
let filteredScripts = []

switch (app.localizedName) {
  case "Code":
    filteredScripts = scripts.filter(
      (s: any) => s.app === "Code"
    )
    break

  case "Google Chrome":
    filteredScripts = scripts.filter(
      (s: any) => s.app === "Google Chrome"
    )
    break
}

let script = await arg<any>("Run", filteredScripts)

await run(script.filePath)
```
![CleanShot 2023-02-14 at 10 44 14](https://user-images.githubusercontent.com/36073/218816342-728d73c3-616a-4888-80dc-56be472972ff.png)

**Comment by yangkennyk** [comment]
> @yangkennyk something like this would do the trick. I'm basing this on custom `// App:` metadata.
> 
> > Not sure of your familiarity with TypeScript, but you can remove the `any` (or if using Typescript, extend/create types)
> 
> ```js
> // Name: Scripts Based on Focused App
> // Shortcut: cmd 0
> 
> import "@johnlindquist/kit"
> 
> let frontmost = await npm("frontmost-app")
> 
> let app = await frontmost()
> 
> let scripts = await getScripts()
> let filteredScripts = []
> 
> switch (app.localizedName) {
>   case "Code":
>     filteredScripts = scripts.filter(
>       (s: any) => s.app === "Code"
>     )
>     break
> 
>   case "Google Chrome":
>     filteredScripts = scripts.filter(
>       (s: any) => s.app === "Google Chrome"
>     )
>     break
> }
> 
> let script = await arg<any>("Run", filteredScripts)
> 
> await run(script.filePath)
> ```
> 
> ![CleanShot 2023-02-14 at 10 44 14](https://user-images.githubusercontent.com/36073/218816342-728d73c3-616a-4888-80dc-56be472972ff.png)

ah yes. This is beautiful. Thank you so much for providing an example!

as an aside. Would it be possible or do you know of anyway to do the following in JavaScript/TypeScript via scriptkit?


https://user-images.githubusercontent.com/6680040/219389695-be26f7fc-78c9-4db9-a9c1-3110fa5560cd.mp4

basically its drawing a green border on the active window to help me easily visually distinguish what window I'm actively focused on because I try to avoid using my mouse as much as I can.

I found out about this via: https://github.com/koekeishiya/limelight and it's now built into https://github.com/koekeishiya/yabai but I've found it buggy at times and resource intensive to use yabai just for the border

I'm currently using a script I wrote in https://www.hammerspoon.org/ to mimic the behavior but it's still buggy at times (I'm also not very knowledgable with lua). Ideally some sort of way to write it in TypeScript would be great.

I've looked into using https://github.com/kasper/phoenix but it lacks the ability to actually draw the border.

thanks again!

**Comment by johnlindquist** [comment]
@yangkennyk Do you want the border on the active window all the time? Or only when Script Kit is open?

I have an idea that might work, I'll have to play around with it though

**Comment by yangkennyk** [comment]
> @yangkennyk Do you want the border on the active window all the time? Or only when Script Kit is open?
> 
> I have an idea that might work, I'll have to play around with it though

a border on the active/focused window at all times would be ideal. Is there anything in script kit right now natively that can  "draw" shapes on the screen?

**Reply by johnlindquist** [reply]
@yangkennyk Here's my attempt:

[Open active-app-frame in Script Kit](https://scriptkit.com/api/new?name=active-app-frame&url=https://gist.githubusercontent.com/johnlindquist/adc83b7748ec8e677e53c8d297f78e54/raw/2354ee0593fec8bb0f2d24e74ca4be94a37fe2fc/active-app-frame.ts")

```js
// Name: Active App Frame

import "@johnlindquist/kit"

let getFrame = async () => {
  // This uses applescript under the hood
  // I'm pretty sure a swift binary would be faster if you want to make one
  let bounds = await getActiveAppBounds()

  let padding = 10
  let x = bounds.left - padding / 2
  let y = bounds.top - padding / 2
  let width = bounds.right - bounds.left + padding
  let height = bounds.bottom - bounds.top + padding

  return { x, y, width, height }
}

let w = await widget(
  `<div class="w-screen h-screen" style="border: 8px solid white">`,
  {
    alwaysOnTop: true,
    ignoreMouse: true,
    transparent: true,
    frame: true,
    ...(await getFrame()),
  }
)

// I don't know of any "app focus change" event to hook into
// Again, it's possible there's a swift solution
setInterval(async () => {
  let { x, y, width, height } = await getFrame()
  w.setPosition(x, y)
  w.setSize(width, height)
}, 250)

// Terminate by either:
// 1. Press cmd+p from the main menu, then select the process to terminate
// 2. Click the menu bar icon, select the process, terminate

```

**Reply by yangkennyk** [reply]
phoenixjs has all the data I would need including window/app change/minimize events: https://kasper.github.io/phoenix/api/events

how could I update the widget outside of script kit?

in phoenixjs I can have it listening for change events which then provides the height/width,x/y I would need for the scripkit widget. 

I'm thinking maybe I could interact with scriptkit from phoenixjs by running scriptkit cli commands within phoenixjs? 

Sounds like it makes sense on paper. I'll tinker around to see if  I can figure something out.

**Reply by johnlindquist** [reply]
@yangkennyk A few ideas (not sure of your experience level, let me know if you'd need any help), in your script your could:

1. include a web server like `express` and send data an endpoint
2. same, but with a unix socket
3. Use chokidar or the `//Watch`  metadata to watch for a file change (does Phoenix store the data somewhere?)

So the question is basically "How would I normally communicate with a node app?"

I would love to see what you come up with if you figure something out. I've never used Phoenix before and it looks really interesting

**Reply by yangkennyk** [reply]
> @yangkennyk A few ideas (not sure of your experience level, let me know if you'd need any help), in your script your could:
> 
> 1. include a web server like `express` and send data an endpoint
> 2. same, but with a unix socket
> 3. Use chokidar or the `//Watch`  metadata to watch for a file change (does Phoenix store the data somewhere?)
> 
> So the question is basically "How would I normally communicate with a node app?"
> 
> I would love to see what you come up with if you figure something out. I've never used Phoenix before and it looks really interesting

it's interesting you say that because creating an express server locally to function as a "command center" for my mac is something I've tinkered around with before but haven't fully fleshed out. I wasn't sure if there was a better solution but you're the first person to actually mention it that I've seen.

basically the local express server would function as an api for all commands/automations on my mac. I've played around with using https://sveinbjorn.org/platypus to create the express "app" and communicating to it for any scripts/automations I want to make. the issue I ran into before is that there are not many tools to interact with my mac via JavaScript before but since then a lot more have popped up.

I've drank the TypeScript kool aid hard. I have the basic understanding down I think. Still working on wrapping my head around generics and writing my own though.

so what I'm thinking is 
- express server   
- + https://trpc.io/ for type safety 
- + https://xstate.js.org/ to manage state
- + scriptkit to interact with express server easily 
- + phoenixjs to manage windows

i could have endpoints like `localhost/trpc/moveWindow` and make a `post` request with whatever args I need etc.

however you lost me at unix socket and chokidar. 

actually I looked up chokidar and it's basically file watching right? so phoenixjs has event listeners 

here's an example of listening to if an app launches for phoenixjs

```
const windowWatcher = Event.on('windowDidFocus', (window) => {
  console.log(window.frame())
});
```

so maybe something like 
```
const windowWatcher = Event.on("windowDidFocus", async (window) => {
  // post to express server that tells scriptkit to update border 
  await fetch("localhost:3000/trpc/updateWindowBorder", {
    method:'POST',
    body: JSON.stringify({ window: window.frame() }),
  });
});
```

however how do I communicate with scriptkit to update the border from the express server? Via the scriptkit cli?

**Reply by johnlindquist** [reply]
Have the route post a JSON object like `{x: 0, y:0. width: 640, height:480}` then pass that data to the widget:


```js
/*
# Command Center
*/

// Name: Command Center
// Background: auto

let w = await widget(
  `<div class="w-screen h-screen" style="border: 8px solid white">`,
  {
    alwaysOnTop: true,
    ignoreMouse: true,
    transparent: true,
  }
)

import "@johnlindquist/kit"

let express = await npm("express")
let port = 5432

let app = express()

app.use(express.json())

app.post("/frame", async (req, res) => {
  let { body } = req

  w.setSize(body.width, body.height)
  w.setPosition(body.x, body.y)

  res.send(body)
})

let url = `http://localhost:${port}`
app.listen(port, () => {
  console.log(`Example app listening at ${url}`)
})

```

Let me know what you come up with!

**Reply by yangkennyk** [reply]
ahhh. you meant making the actual widget I'm creating in script kit an express app. That makes sense. I'll let you know what I come up with!

**Comment by yangkennyk** [comment]
@johnlindquist so I have a working configuration however I'm noticing sometimes I'm getting this weird like "shadow border". It seems random. tried to adjust the widget settings to see what could be causing it but unsure. 

any ideas?

![image](https://user-images.githubusercontent.com/6680040/225423716-cdb19a4a-a6c0-4a7b-b08f-f95c99c207bf.png)

**Reply by johnlindquist** [reply]
@yangkennyk Try:

`hasShadow: false,`

in the widget settings

**Comment by yangkennyk** [comment]
@johnlindquist thanks! Not sure how I didn't see that option. 

wasn't sure if I should open a ticket for this question. But is there a way to terminate the process via shortcuts? 

Right now I'm currently

1. save the script
2. test it
3. something isn't right.
4. mouse over menu item to "running processes"
5. click terminate
6. start the script again.

I looked through the docs but wasn't able to find a way to terminate the script in a faster way. Would it be possible to automatically rerun the script on save?

**Reply by johnlindquist** [reply]
@yangkennyk 

```
// Background: auto
```

Will auto re-run on save.

Otherwise, `cmd+shift+;` _should_ end the most recently run script. It's not super-well tested though.
